ctapipe/io/dl1eventsource.py

```diff
@@ -0,0 +1,427 @@
+                        tel.name
+                    ].iterrows()
+                    for tel in self.file_.root.simulation.event.telescope.images
+                }
+
+        if DataLevel.DL1_PARAMETERS in self.datalevels:
+            param_readers = {
+                tel.name: HDF5TableReader(self.file_).read(
+                    f"/dl1/event/telescope/parameters/{tel.name}",
+                    containers=[
+                        HillasParametersContainer(),
+                        TimingParametersContainer(),
+                        LeakageContainer(),
+                        ConcentrationContainer(),
+                        MorphologyContainer(),
+                        IntensityStatisticsContainer(),
+                        PeakTimeStatisticsContainer(),
+                    ],
+                    prefixes=True,
+                )
+                for tel in self.file_.root.dl1.event.telescope.parameters
+            }
+            if self.has_simulated_dl1:
+                simulated_param_readers = {
+                    tel.name: HDF5TableReader(self.file_).read(
+                        f"/simulation/event/telescope/parameters/{tel.name}",
+                        containers=[
+                            HillasParametersContainer(),
+                            TimingParametersContainer(),
+                            LeakageContainer(),
+                            ConcentrationContainer(),
+                            MorphologyContainer(),
+                            IntensityStatisticsContainer(),
+                            PeakTimeStatisticsContainer(),
+                        ],
+                        prefixes=True,
+                    )
+                    for tel in self.file_.root.dl1.event.telescope.parameters
+                }
+
+        if self.is_simulation:
+            # simulated shower wide information
+            mc_shower_reader = HDF5TableReader(self.file_).read(
+                "/simulation/event/subarray/shower",
+                SimulatedShowerContainer(),
+                prefixes="true",
+            )
+
+        # Setup iterators for the array events
+        events = HDF5TableReader(self.file_).read(
+            "/dl1/event/subarray/trigger", [TriggerContainer(), EventIndexContainer()]
+        )
+
+        array_pointing_finder = IndexFinder(
+            self.file_.root.dl1.monitoring.subarray.pointing.col("time")
+        )
+
+        tel_pointing_finder = {
+            tel.name: IndexFinder(tel.col("telescopetrigger_time"))
+            for tel in self.file_.root.dl1.monitoring.telescope.pointing
+        }
+
+        for counter, array_event in enumerate(events):
+            data.dl1.tel.clear()
+            data.simulation.tel.clear()
+            data.pointing.tel.clear()
+            data.trigger.tel.clear()
+
+            data.count = counter
+            data.trigger, data.index = next(events)
+            data.trigger.tels_with_trigger = (
```

Comment:
Add this to the subarray? We have a method id list → mask, so it would only make sense to also implement the other way around there.

Reply:
You mean adding a new, general method to the subarray class and use it here?

Reply:
yes